src/openssl/ssl/context.cr

```diff
@@ -125,60 +125,48 @@ abstract class OpenSSL::SSL::Context
     # context = OpenSSL::SSL::Context::Server.new
     # context.add_options(OpenSSL::SSL::Options::NO_SSL_V2 | OpenSSL::SSL::Options::NO_SSL_V3)
     # ```
-    def initialize(method : LibSSL::SSLMethod = Context.default_method)
-      super(method)
-
-      add_options(OpenSSL::SSL::Options::CIPHER_SERVER_PREFERENCE)
+    def self.new(method : LibSSL::SSLMethod = Context.default_method)
+      context = new(method, secure: true)
+      context.add_options(OpenSSL::SSL::Options::CIPHER_SERVER_PREFERENCE)
       {% if LibSSL::OPENSSL_102 %}
-      self.default_verify_param = "ssl_server"
+      context.default_verify_param = "ssl_server"
       {% end %}
+      context
     end
 
     # Returns a new TLS server context with only the given method set.
     #
     # For everything else this uses the defaults of your OpenSSL.
     # Use this only if undoing the defaults that `new` sets is too much hassle.
     def self.insecure(method : LibSSL::SSLMethod = Context.default_method)
-      super(method)
+      new method, secure: false
     end
   end
 
-  protected def initialize(method : LibSSL::SSLMethod)
-    @handle = LibSSL.ssl_ctx_new(method)
-    raise OpenSSL::Error.new("SSL_CTX_new") if @handle.null?
-
-    set_default_verify_paths
-
-    add_options(OpenSSL::SSL::Options.flags(
-      ALL,
-      NO_SSL_V2,
-      NO_SSL_V3,
-      NO_SESSION_RESUMPTION_ON_RENEGOTIATION,
-      SINGLE_ECDH_USE,
-      SINGLE_DH_USE
-    ))
-
-    add_modes(OpenSSL::SSL::Modes.flags(AUTO_RETRY, RELEASE_BUFFERS))
-
-    self.ciphers = CIPHERS
-
-    set_tmp_ecdh_key(curve: LibCrypto::NID_X9_62_prime256v1)
-  end
+  protected def self.new(method : LibSSL::SSLMethod, secure : Bool)
+    handle = LibSSL.ssl_ctx_new(method)
+    raise OpenSSL::Error.new("SSL_CTX_new") if handle.null?
+    context = new(handle)
+
+    if secure
```

Comment:
Wouldn't it be better to extract this to an additional protected constructor `new_secure` instead of an argument? The insecure constructors call protected `new` directly, secure constructors through `new_secure`.